kaggle_environments/envs/halite/helpers.py

```diff
@@ -0,0 +1,781 @@
+                    else None
+                )
+                self._add_shipyard(Shipyard(shipyard_id, shipyard_position, player_id, self, action))
+
+    @property
+    def configuration(self) -> Configuration:
+        return self._configuration
+
+    @property
+    def players(self) -> Dict[PlayerId, Player]:
+        return self._players
+
+    @property
+    def ships(self) -> Dict[ShipId, Ship]:
+        """Returns all ships on the current board."""
+        return self._ships
+
+    @property
+    def shipyards(self) -> Dict[ShipyardId, Shipyard]:
+        """Returns all shipyards on the current board."""
+        return self._shipyards
+
+    @property
+    def cells(self) -> Dict[Point, Cell]:
+        """Returns all cells on the current board."""
+        return self._cells
+
+    @property
+    def step(self) -> int:
+        return self._step
+
+    @property
+    def current_player_id(self) -> PlayerId:
+        return self._current_player_id
+
+    @property
+    def current_player(self) -> Player:
+        """Returns the current player (generally this is you)."""
+        return self._players[self.current_player_id]
+
+    @property
+    def opponents(self) -> List[Player]:
+        """
+        Returns all players that aren't the current player.
+        You can get all opponent ships with [ship for ship in player.ships for player in board.opponents]
+        """
+        return [player for player in self.players.values() if not player.is_current_player]
+
+    @property
+    def observation(self) -> Dict[str, Any]:
+        """Converts a Board back to the normalized observation that constructed it."""
+        size = self.configuration.size
+        halite = [self[index_to_position(index, size)].halite for index in range(size * size)]
+        players = [player.observation for player in self.players.values()]
+
+        return {
+            "halite": halite,
+            "players": players,
+            "player": self.current_player_id,
+            "step": self.step
+        }
+
+    def __deepcopy__(self, _) -> 'Board':
+        actions = [player.next_actions for player in self.players.values()]
+        return Board(self.observation, self.configuration, actions)
+
+    def __getitem__(self, position: Point) -> Cell:
+        """
+        This method will wrap the supplied position to fit within the board size and return the cell at that location.
+        e.g. on a 3x3 board, board[(2, 1)] is the same as board[(5, 1)]
+        """
+        return self._cells[wrap_point(position, self.configuration.size)]
+
+    def __str__(self) -> str:
+        """
+        The board is printed in a grid with the following rules:
+        Capital letters are shipyards
+        Lower case letters are ships
+        Digits are cell halite and scale from 0-9 directly proportional to a value between 0 and self.configuration.max_cell_halite
+        Player 1 is letter a/A
+        Player 2 is letter b/B
+        etc.
+        """
```

Comment:
I'm curious if people think this belongs in `__str__` or another method like `print_board` or something like that?